gsd/gsd.c

```diff
@@ -1187,41 +2039,73 @@ gsd_find_chunk(struct gsd_handle* handle, uint64_t frame, const char* name)
     }
 
     // find the id for the given name
-    uint16_t match_id = gsd_get_id(handle, name, 0);
+    uint16_t match_id = gsd_name_id_map_find(&handle->name_map, name);
     if (match_id == UINT16_MAX)
     {
         return NULL;
     }
 
-    // binary search for the first index entry at the requested frame
-    size_t L = 0;
-    size_t R = handle->index_num_entries;
-
-    // progressively narrow the search window by halves
-    do
+    if (handle->header.gsd_version >= gsd_make_version(2, 0))
     {
-        size_t m = (L + R) / 2;
+        // gsd 2.0 files sort the entire index
+        // binary search for the index entry
+        ssize_t L = 0;
+        ssize_t R = handle->file_index.size - 1;
+        struct gsd_index_entry T;
+        T.frame = frame;
+        T.id = match_id;
 
-        if (frame < handle->index[m].frame)
+        while (L <= R)
         {
-            R = m;
+            size_t m = (L + R) / 2;
+            int cmp = gsd_cmp_index_entry(handle->file_index.data + m, &T);
+            if (cmp == -1)
+            {
+                L = m + 1;
+            }
+            else if (cmp == 1)
+            {
+                R = m - 1;
+            }
+            else
+            {
+                return &(handle->file_index.data[m]);
+            }
```

Comment:
Is it worth putting this into a `gsd_heap_search` function or so?

Reply:
Maybe, though it is only used once in this function. We can certainly factor it out if we ever need to use it in more than one place.